File: apps/worker/src/tcabr_worker/exclusions.py
from __future__ import annotations

import asyncpg


async def load_excluded(pool: asyncpg.Pool, usernames: list[str]) -> set[str]:
    if not usernames:
        return set()
    async with pool.acquire() as c:
        rows = await c.fetch(
            "select gh_username from user_exclusion where gh_username = any($1::text[])",
            usernames,
        )
    return {r["gh_username"] for r in rows}

File: apps/web/lib/exclusions.ts
import { supabaseService } from "./supabase";

export async function loadExcluded(usernames: string[]): Promise<Set<string>> {
  if (usernames.length === 0) return new Set();
  const db = supabaseService();
  const { data } = await db
    .from("user_exclusion")
    .select("gh_username")
    .in("gh_username", usernames);
  return new Set((data ?? []).map((r) => r.gh_username));
}

export function filterExcluded<T>(
  rows: T[],
  excluded: Set<string>,
  getUsername: (row: T) => string,
): T[] {
  return rows.filter((r) => !excluded.has(getUsername(r)));
}
